lib/window-size-counter.c

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2002-2018 Balabit
+ * Copyright (c) 2018 Laszlo Budai <laszlo.budai@balabit.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "window-size-counter.h"
+
+#define COUNTER_MASK (((gsize)1<<(8*sizeof(gsize)-1)) - 1)
+static const gsize COUNTER_MAX = COUNTER_MASK;
+static const gsize SUSPEND_MASK = G_MAXSIZE ^ COUNTER_MASK;
+static const gsize RESUME_MASK = G_MAXSIZE >> 1;
```

Comment:
`atomic_gssize` is implemented with atomic pointers, so I would add a static_assert to compare its size with gsize if we want to use G_MAXSIZE on it.